[CoordinatedGraphics] Tiled backing store is not needed in paintTile()

carlosgcampos opened the pull request:
#### 58234aa593e4507d3a697a4b5ed2fc00b0e2e5e7
<pre>
[CoordinatedGraphics] Tiled backing store is not needed in paintTile()
<a href="https://bugs.webkit.org/show_bug.cgi?id=280679">https://bugs.webkit.org/show_bug.cgi?id=280679</a>

Reviewed by Nikolas Zimmermann.

The function to map the dirty rect to contents is just scaling the
rectangle, and scale is already available by effectiveContentsScale().
Also make paintIntoGraphicsContext Skia specific, since it&apos;s only used
my Skia, and move it to CoordinatedGraphicsLayerSkia.cpp.

* Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp:
(WebCore::SkiaThreadedPaintingPool::recordDisplayList const):
(WebCore::SkiaThreadedPaintingPool::postPaintingTask):
* Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
(WebCore::CoordinatedGraphicsLayer::effectiveContentsScale const):
(WebCore::CoordinatedGraphicsLayer::updateContentBuffers):
(WebCore::CoordinatedGraphicsLayer::selfOrAncestorHaveNonAffineTransforms const):
(WebCore::CoordinatedGraphicsLayer::effectiveContentsScale): Deleted.
(WebCore::CoordinatedGraphicsLayer::paintIntoGraphicsContext const): Deleted.
(WebCore::CoordinatedGraphicsLayer::selfOrAncestorHaveNonAffineTransforms): Deleted.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerCairo.cpp:
(WebCore::CoordinatedGraphicsLayer::paintTile):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp:
(WebCore::CoordinatedGraphicsLayer::paintIntoGraphicsContext const):
(WebCore::CoordinatedGraphicsLayer::paintTile):

Canonical link: <a href="https://commits.webkit.org/284544@main">https://commits.webkit.org/284544@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/607af8be9f8f36b96b5aa4b64edad6ce3384a6a7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69799 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49200 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22552 "Build is in progress. Recent messages:OS: Ventura (13.6.7), Xcode: 14.3; Skipping applying patch since patch_id isn't provided; Running checkout-pull-request; Compiled WebKit (warnings)") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/73884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/20957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71916 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57000 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20808 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/73884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/20957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72865 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/44841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/60201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/73884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/41506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/17641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/19334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/63435 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/17985 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75598 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/87/builds/14024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/17223 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63111 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/14059 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/60282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/63038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/4661 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10647 "Built successfully and passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45003 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Running apply-patch; Checked out pull request; Compiled WebKit (warnings)") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/46077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/47348 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45818 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp:
```diff
@@ -47,7 +47,27 @@
+
+    context.translate(-dirtyRect.x(), -dirtyRect.y());
+    context.scale(scale);
+    paintGraphicsLayerContents(context, clipRect);
```
You are passing in the non-enclosed version of clipRect, whereas we previously passed the rect after mapToContents(), which calls enclosingIntRect. I guess that behavior change wasn't intentional?

carlosgcampos replied:
Yes, it is intentional, because paintGraphicsLayerContents receives a FloatRect, so there's no point in converting to IntRect, no?

nikolaszimmermann replied:
The old code used `paintGraphicsLayerContents(context, tiledBackingStore.mapToContents(dirtyRect));`, and `mapToContents()` returned `enclosingIntRect(x / scale, y / scale, w / scale, h / scale)`. Now you're passing `(x / scale, y / scale, w / scale, h / scale)` here without taking the enclosingIntRect.

paintGraphicsLayerContents allows clipping on non-integer boundaries, but we previously didn't make use of that -- therefore it's a subtle change in behavior, and we should be sure that it's correct.

carlosgcampos replied:
Ok, let's not change the behavior in a refactoring patch.

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerCairo.cpp:
```diff
@@ -21,17 +21,21 @@
+
     auto buffer = Nicosia::UnacceleratedBuffer::create(dirtyRect.size(), contentsOpaque() ? Nicosia::Buffer::NoFlags : Nicosia::Buffer::SupportsAlpha);
-    m_coordinator->paintingEngine().paint(*this, buffer.get(), dirtyRect, mappedDirtyRect, IntRect { { 0, 0 }, dirtyRect.size() }, tiledBackingStore.contentsScale());
+    m_coordinator->paintingEngine().paint(*this, buffer.get(), dirtyRect, enclosingIntRect(scaledDirtyRect), IntRect { { 0, 0 }, dirtyRect.size() }, scale);
```
s/ { 0, 0 } / { } /

nikolaszimmermann reviewed (approved):
r=me, with a comment:

carlosgcampos committed:
[CoordinatedGraphics] Tiled backing store is not needed in paintTile()
https://bugs.webkit.org/show_bug.cgi?id=280679

Reviewed by Nikolas Zimmermann.

The function to map the dirty rect to contents is just scaling the
rectangle, and scale is already available by effectiveContentsScale().
Also make paintIntoGraphicsContext Skia specific, since it's only used
my Skia, and move it to CoordinatedGraphicsLayerSkia.cpp.

* Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp:
(WebCore::SkiaThreadedPaintingPool::recordDisplayList const):
(WebCore::SkiaThreadedPaintingPool::postPaintingTask):
* Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
(WebCore::CoordinatedGraphicsLayer::effectiveContentsScale const):
(WebCore::CoordinatedGraphicsLayer::updateContentBuffers):
(WebCore::CoordinatedGraphicsLayer::selfOrAncestorHaveNonAffineTransforms const):
(WebCore::CoordinatedGraphicsLayer::effectiveContentsScale): Deleted.
(WebCore::CoordinatedGraphicsLayer::paintIntoGraphicsContext const): Deleted.
(WebCore::CoordinatedGraphicsLayer::selfOrAncestorHaveNonAffineTransforms): Deleted.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerCairo.cpp:
(WebCore::CoordinatedGraphicsLayer::paintTile):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp:
(WebCore::CoordinatedGraphicsLayer::paintIntoGraphicsContext const):
(WebCore::CoordinatedGraphicsLayer::paintTile):

Canonical link: https://commits.webkit.org/284544@main
